### PR TITLE
Add reshaping for distributions.

### PR DESCRIPTION
**Pull request by tillahoffmann**
In light of #76709, I've been experimenting with reshaping for distributions. There seems to be a generic-ish way of handling reshaping. It comprises two parts:

1. A property `_reshape_args` (questionable name) that yields the names of arguments that need to be reshaped. It defaults to `arg_constraints` such that distributions like `Beta`, `Normal`, `Gamma` (and others with unambiguous arguments) can be handled in a generic way. For distributions with ambiguous arguments (such as `Bernoulli` which takes `probs` or `logits`), we yield the name of whichever argument was used to create the class.
2. An implementation in the `Distribution` base class that fetches all arguments that need to be reshaped, reshapes them, and creates a new instance with `validate_args=False` to skip the validation step.

This PR is very much a draft and requires further discussion.

### CODE REVIEW

File: torch/distributions/relaxed_bernoulli.py [hunk 1]
```diff
@@ -47,6 +48,11 @@ def __init__(self, temperature, probs=None, logits=None, validate_args=None):
             batch_shape = self._param.size()
         super(LogitRelaxedBernoulli, self).__init__(batch_shape, validate_args=validate_args)
 
+    @property
+    def _reshape_args(self):
+        yield 'temperature', False
```
Comment (tillahoffmann) [inline torch/distributions/relaxed_bernoulli.py:53]: `temperature` doesn't appear to be included in `broadcast_all` in the constructor, so I've omitted it from reshaping here.
